File: splitbill-llmocr-web/src/lib/services/billService.ts
import axios from 'axios';

export interface CreateBillPayload {
  name: string;
  tax_amount: number;
  tip_amount: number;
}

export interface Bill {
  id: string;
  name: string;
  tax_amount: number;
  tip_amount: number;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface BillWithItems extends Bill {
  items?: BillItem[];
  participants?: BillParticipant[];
}

export interface BillItem {
  id: number;
  bill_id: string;
  name: string;
  price: number;
  quantity: number;
  created_at: string;
}

export interface BillParticipant {
  id: number;
  bill_id: string;
  name: string;
  payment_status: string;
  share_of_common_costs: number;
  created_at: string;
}

export interface BillStatus {
  bill_id: string;
  status: string;
}

export interface UploadImageResponse {
  message: string;
  image_url?: string;
}

export interface ItemAssignment {
  item_id: number;
  participant_id: number;
}

export interface UpdateItemPayload {
  name?: string;
  price?: number;
  quantity?: number;
}

export interface UpdateBillPayload {
  tax_amount?: number;
  tip_amount?: number;
}

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8080';

export const billService = {
  async createBill(payload: CreateBillPayload): Promise<Bill> {
    try {
      const response = await axios.post(`${API_BASE_URL}/api/bills/`, payload, {
        headers: {
          'Content-Type': 'application/json',
        },
      });
      return response.data as Bill;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to create bill');
    }
  },

  async getBills(): Promise<Bill[]> {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/bills`);
      return response.data as Bill[];
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to fetch bills');
    }
  },

  async getBillById(id: string): Promise<Bill> {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/bills/${id}`);
      return response.data as Bill;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to fetch bill');
    }
  },

  async uploadBillImage(billId: string, imageFile: File): Promise<UploadImageResponse> {
    try {
      const formData = new FormData();
      formData.append('image', imageFile);
      
      const response = await axios.post(`${API_BASE_URL}/api/bills/${billId}/image`, formData, {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      });
      return response.data as UploadImageResponse;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to upload image');
    }
  },

  async getBillStatus(billId: string): Promise<BillStatus> {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/bills/${billId}/status`);
      return response.data as BillStatus;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to fetch bill status');
    }
  },

  async getBillWithItems(billId: string): Promise<BillWithItems> {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/bills/${billId}`);
      return response.data as BillWithItems;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to fetch bill with items');
    }
  },

  async getParticipants(billId: string): Promise<BillParticipant[]> {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/bills/${billId}/participants`);
      return response.data as BillParticipant[];
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to fetch participants');
    }
  },

  async getItemAssignments(billId: string): Promise<ItemAssignment[]> {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/bills/${billId}/item-assignments`);
      
      if (response.data && Array.isArray(response.data)) {
        return response.data.map((assignment: { item_id: number; participant_id: number }) => ({
          item_id: assignment.item_id,
          participant_id: assignment.participant_id
        }));
      }
      
      return [];
    } catch (error) {
      console.error('Error fetching item assignments:', error);
      return [];
    }
  },

  async addParticipant(billId: string, participant: Omit<BillParticipant, 'id' | 'bill_id' | 'created_at'>): Promise<BillParticipant> {
    try {
      const response = await axios.post(`${API_BASE_URL}/api/bills/${billId}/participants`, participant, {
        headers: {
          'Content-Type': 'application/json',
        },
      });
      return response.data as BillParticipant;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to add participant');
    }
  },

  async assignItemToParticipant(billId: string, itemId: number, participantId: number): Promise<ItemAssignment> {
    try {
      const response = await axios.post(`${API_BASE_URL}/api/bills/${billId}/assign-items`, {
        item_id: itemId,
        participant_id: participantId,
      }, {
        headers: {
          'Content-Type': 'application/json',
        },
      });
      return response.data as ItemAssignment;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to assign item to participant');
    }
  },

  async deleteItemAssignment(billId: string, itemId: number, participantId: number): Promise<void> {
    try {
      await axios({
        method: 'delete',
        url: `${API_BASE_URL}/api/bills/${billId}/assign-items`,
        data: {
          item_id: itemId,
          participant_id: participantId,
        },
        headers: {
          'Content-Type': 'application/json',
        },
      });
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to remove item assignment');
    }
  },

  async deleteParticipant(billId: string, participantId: number): Promise<void> {
    try {
      await axios.delete(`${API_BASE_URL}/api/bills/${billId}/participants/${participantId}`);
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to delete participant');
    }
  },

  async updateItem(itemId: number, updates: UpdateItemPayload): Promise<BillItem> {
    try {
      const response = await axios.put(`${API_BASE_URL}/api/items/${itemId}`, updates, {
        headers: {
          'Content-Type': 'application/json',
        },
      });
      return response.data as BillItem;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to update item');
    }
  },

  async updateBill(billId: string, updates: UpdateBillPayload): Promise<Bill> {
    try {
      const response = await axios.put(`${API_BASE_URL}/api/bills/${billId}`, updates, {
        headers: {
          'Content-Type': 'application/json',
        },
      });
      return response.data as Bill;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to update bill');
    }
  },

  async getBillSummary(billId: string): Promise<unknown> {
    try {
      const response = await axios.get(`${API_BASE_URL}/api/bills/${billId}/summary`);
      return response.data;
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error && error.response && typeof error.response === 'object' && 'data' in error.response && error.response.data && typeof error.response.data === 'object' && 'message' in error.response.data) {
        throw new Error((error.response.data as { message: string }).message);
      }
      throw new Error('Failed to fetch bill summary');
    }
  },
};

File: splitbill-llmocr-api/internal/services/bill_service.go
package services

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"time"

	"github.com/Aebroyx/splitbill-llmocr-api/internal/domain/models"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

type BillService struct {
	db *gorm.DB
}

func NewBillService(db *gorm.DB) *BillService {
	return &BillService{db: db}
}

// GetDB returns the database instance
func (s *BillService) GetDB() *gorm.DB {
	return s.db
}

// CreateBill creates a new bill
func (s *BillService) CreateBill(req *models.BillRequest) (*models.BillResponse, error) {
	bill := &models.Bills{
		ID:        uuid.New(),
		Name:      req.Name,
		Status:    "active",
		TaxAmount: req.TaxAmount,
		TipAmount: req.TipAmount,
	}

	if err := s.db.Create(bill).Error; err != nil {
		return nil, fmt.Errorf("failed to create bill: %w", err)
	}

	return s.getBillResponse(bill), nil
}

// GetBill retrieves a bill by ID
func (s *BillService) GetBill(id uuid.UUID) (*models.BillResponse, error) {
	var bill models.Bills
	if err := s.db.Preload("Items").Preload("Participants").First(&bill, "id = ?", id).Error; err != nil {
		return nil, fmt.Errorf("bill not found: %w", err)
	}

	return s.getBillResponse(&bill), nil
}

// UploadBillImage uploads an image for a bill and triggers n8n workflow
func (s *BillService) UploadBillImage(billID uuid.UUID, file *multipart.FileHeader) (*models.BillResponse, error) {
	// Check if bill exists
	bill, err := s.GetBill(billID)
	if err != nil {
		return nil, fmt.Errorf("bill not found: %w", err)
	}

	// Read file data
	fileBytes, err := s.readFileData(file)
	if err != nil {
		return nil, fmt.Errorf("failed to read file data: %w", err)
	}

	// Save image to disk (optional, for backup)
	imagePath := fmt.Sprintf("./uploads/bill_%s_%s", billID.String(), file.Filename)
	if err := os.MkdirAll("./uploads", 0755); err != nil {
		fmt.Printf("Failed to create uploads directory: %v\n", err)
		// Don't fail the upload for this, continue with n8n
	}

	if err := os.WriteFile(imagePath, fileBytes, 0644); err != nil {
		fmt.Printf("Failed to save image to disk: %v\n", err)
		// Don't fail the upload for this, continue with n8n
	}

	// Trigger n8n workflow with image data
	if err := s.triggerN8nWorkflowWithImage(billID, fileBytes, file.Filename); err != nil {
		// If n8n workflow fails, the status should already be set to "failed"
		// but let's make sure we return a proper error message
		fmt.Printf("N8n workflow failed for bill %s: %v\n", billID, err)
		return nil, fmt.Errorf("failed to process image with AI: %w", err)
	}

	return bill, nil
}

// readFileData reads the file data from multipart.FileHeader into bytes
func (s *BillService) readFileData(file *multipart.FileHeader) ([]byte, error) {
	src, err := file.Open()
	if err != nil {
		return nil, fmt.Errorf("failed to open uploaded file: %w", err)
	}
	defer src.Close()

	// Read file content into bytes
	fileBytes, err := io.ReadAll(src)
	if err != nil {
		return nil, fmt.Errorf("failed to read file: %w", err)
	}

	return fileBytes, nil
}

// triggerN8nWorkflowWithImage sends the image data directly to n8n workflow
func (s *BillService) triggerN8nWorkflowWithImage(billID uuid.UUID, imageData []byte, filename string) error {
	n8nWebhookURL := os.Getenv("N8N_WEBHOOK_URL")
	if n8nWebhookURL == "" {
		err := fmt.Errorf("N8N_WEBHOOK_URL not configured")
		fmt.Printf("N8N_WEBHOOK_URL not configured, skipping workflow trigger for bill %s\n", billID)
		// Update bill status to failed since we can't process
		if updateErr := s.UpdateBillStatus(billID, "failed"); updateErr != nil {
			fmt.Printf("Failed to update bill status to failed: %v\n", updateErr)
		}
		return err
	}

	// Create multipart form data
	var requestBody bytes.Buffer
	writer := multipart.NewWriter(&requestBody)

	// Add bill_id field
	if err := writer.WriteField("bill_id", billID.String()); err != nil {
		fmt.Printf("Failed to write bill_id field: %v\n", err)
		// Update bill status to failed
		if updateErr := s.UpdateBillStatus(billID, "failed"); updateErr != nil {
			fmt.Printf("Failed to update bill status to failed: %v\n", updateErr)
		}
		return fmt.Errorf("failed to write bill_id field: %v", err)
	}

	// Add image file
	part, err := writer.CreateFormFile("image", filename)
	if err != nil {
		fmt.Printf("Failed to create form file: %v\n", err)
		// Update bill status to failed
		if updateErr := s.UpdateBillStatus(billID, "failed"); updateErr != nil {
			fmt.Printf("Failed to update bill status to failed: %v\n", updateErr)
		}
		return fmt.Errorf("failed to create form file: %v", err)
	}
	if _, err := part.Write(imageData); err != nil {
		fmt.Printf("Failed to write image data: %v\n", err)
		// Update bill status to failed
		if updateErr := s.UpdateBillStatus(billID, "failed"); updateErr != nil {
			fmt.Printf("Failed to update bill status to failed: %v\n", updateErr)
		}
		return fmt.Errorf("failed to write image data: %v", err)
	}

	// Get the Content-Type BEFORE closing the writer
	contentType := writer.FormDataContentType()

	// Close the writer to finalize the multipart data
	writer.Close()

	// Send request to n8n
	req, err := http.NewRequest("POST", n8nWebhookURL, &requestBody)
	if err != nil {
		fmt.Printf("Failed to create request: %v\n", err)
		// Update bill status to failed
		if updateErr := s.UpdateBillStatus(billID, "failed"); updateErr != nil {
			fmt.Printf("Failed to update bill status to failed: %v\n", updateErr)
		}
		return fmt.Errorf("failed to create request: %v", err)
	}

	// Set the Content-Type header with the boundary
	req.Header.Set("Content-Type", contentType)

	// Set timeout for the request
	client := &http.Client{
		Timeout: 30 * time.Second, // 30 second timeout
	}

	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("Failed to send request to n8n: %v\n", err)
		// Update bill status to failed
		if updateErr := s.UpdateBillStatus(billID, "failed"); updateErr != nil {
			fmt.Printf("Failed to update bill status to failed: %v\n", updateErr)
		}
		return fmt.Errorf("failed to send request to n8n: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		bodyBytes, _ := io.ReadAll(resp.Body)
		fmt.Printf("N8n workflow returned status: %d\n", resp.StatusCode)
		fmt.Printf("Response body: %s\n", string(bodyBytes))
		fmt.Printf("Request headers: %v\n", req.Header)

		// Update bill status to failed since n8n workflow failed
		if updateErr := s.UpdateBillStatus(billID, "failed"); updateErr != nil {
			fmt.Printf("Failed to update bill status to failed: %v\n", updateErr)
		}

		return fmt.Errorf("n8n workflow failed with status %d: %s", resp.StatusCode, string(bodyBytes))
	}

	fmt.Printf("Successfully triggered n8n workflow for bill %s\n", billID)
	return nil
}

// ProcessExtractedData processes the data returned from n8n workflow
func (s *BillService) ProcessExtractedData(billID uuid.UUID, extractedData string) error {
	var bill models.Bills
	if err := s.db.First(&bill, "id = ?", billID).Error; err != nil {
		return fmt.Errorf("bill not found: %w", err)
	}

	// Parse the extracted data
	var extractedItems models.ExtractedItemData
	if err := json.Unmarshal([]byte(extractedData), &extractedItems); err != nil {
		fmt.Printf("Failed to parse JSON: %v\n", err)
		return fmt.Errorf("failed to parse extracted data: %w", err)
	}

	// Start a transaction
	tx := s.db.Begin()
	defer func() {
		if r := recover(); r != nil {
			tx.Rollback()
		}
	}()

	// Update bill with extracted data (only tax and tip amounts)
	if err := tx.Model(&bill).Updates(map[string]interface{}{
		"tax_amount": extractedItems.Tax,
		"tip_amount": extractedItems.Tip,
	}).Error; err != nil {
		tx.Rollback()
		return fmt.Errorf("failed to update bill: %w", err)
	}

	// Create items from extracted data
	for _, item := range extractedItems.Items {
		dbItem := models.Items{
			BillID:   billID,
			Name:     item.Name,
			Price:    item.Price,
			Quantity: item.Quantity,
		}

		if err := tx.Create(&dbItem).Error; err != nil {
			tx.Rollback()
			return fmt.Errorf("failed to create item: %w", err)
		}
	}

	return tx.Commit().Error
}

// GetBillSummary calculates and returns bill summary
func (s *BillService) GetBillSummary(billID uuid.UUID) (*models.BillSummary, error) {
	var bill models.Bills
	if err := s.db.Preload("Items").Preload("Participants").First(&bill, "id = ?", billID).Error; err != nil {
		return nil, fmt.Errorf("bill not found: %w", err)
	}

	// Calculate total items
	var totalItems float64
	for _, item := range bill.Items {
		totalItems += item.Price * float64(item.Quantity)
	}

	// Calculate participant shares
	participantShares := make(map[string]float64)
	totalParticipants := len(bill.Participants)
	if totalParticipants > 0 {
		sharePerPerson := (totalItems + bill.TaxAmount + bill.TipAmount) / float64(totalParticipants)
		for _, participant := range bill.Participants {
			participantShares[participant.Name] = sharePerPerson + participant.ShareOfCommonCosts
		}
	}

	return &models.BillSummary{
		BillID:            billID,
		TotalItems:        totalItems,
		TaxAmount:         bill.TaxAmount,
		TipAmount:         bill.TipAmount,
		TotalBill:         totalItems + bill.TaxAmount + bill.TipAmount,
		ParticipantShares: participantShares,
	}, nil
}

// UpdateBillStatus updates the status of a bill
func (s *BillService) UpdateBillStatus(billID uuid.UUID, status string) error {
	return s.db.Model(&models.Bills{}).Where("id = ?", billID).Update("status", status).Error
}

// GetBillStatus returns the current status of a bill
func (s *BillService) GetBillStatus(billID uuid.UUID) (string, error) {
	var bill models.Bills
	err := s.db.Select("status").Where("id = ?", billID).First(&bill).Error
	if err != nil {
		return "", err
	}
	return bill.Status, nil
}

// getBillResponse converts a Bills model to BillResponse
func (s *BillService) getBillResponse(bill *models.Bills) *models.BillResponse {
	response := &models.BillResponse{
		ID:        bill.ID,
		Name:      bill.Name,
		Status:    bill.Status,
		TaxAmount: bill.TaxAmount,
		TipAmount: bill.TipAmount,
		CreatedAt: bill.CreatedAt,
	}

	// Convert items
	for _, item := range bill.Items {
		response.Items = append(response.Items, models.ItemResponse{
			ID:        item.ID,
			BillID:    item.BillID,
			Name:      item.Name,
			Price:     item.Price,
			Quantity:  item.Quantity,
			CreatedAt: item.CreatedAt,
		})
	}

	// Convert participants
	for _, participant := range bill.Participants {
		response.Participants = append(response.Participants, models.ParticipantResponse{
			ID:                 participant.ID,
			BillID:             participant.BillID,
			Name:               participant.Name,
			PaymentStatus:      participant.PaymentStatus,
			ShareOfCommonCosts: participant.ShareOfCommonCosts,
			CreatedAt:          participant.CreatedAt,
		})
	}

	return response
}
